### PR TITLE
Add ability to specify SOCKS proxy for SSH connections

### LINKED ISSUE #1539: Add SOCKS5 SSH proxy support in Ray
Related to #1538, enables running Modin with Ray autoscaler in a typical enterprise network.

### PR DESCRIPTION
**Pull request by vnlitvinov**
## Why are these changes needed?
In some cases like corporate networks external SSH connections must go through a proxy.
This change adds optional field in the autoscaler configuration to specify socks host (and port) for SSH. In current implementation it requires that `nc` is also installed, as it is used internally as `ProxyCommand`.

## Related issue number
Closes https://github.com/modin-project/modin/issues/1539
I know this is an external issue, I can make an issue inside Ray project if needed.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [x] This PR is not tested (please justify below)

## Testing strategy justification

There is currently no automated testing for this, but I have added some validation in JSON schema, and we verified this works correctly manually. I'm open at adding tests if pointed at best location for those, as I don't know Ray codebase well enough yet.

### CODE REVIEW

**Comment by vnlitvinov** [comment]
How do I reproduce the CI failures on Windows? Any docs on setting up the machine?

**Comment by rkooo567** [comment]
cc @mehrdadn 
cc @thomasdesr

File: python/ray/autoscaler/updater.py
Comment (mehrdadn) [inline python/ray/autoscaler/updater.py:281]: Try using `shlex.join` here instead. (It handles quoting each argument, so you can avoid extra manual quoting.)

**Comment by thomasdesr** [comment]
Hey @vnlitvinov! Thanks for sending the PR; something like this definitely makes sense to add :D

If you're currently blocked on this, you should be able to unblock by setting `ProxyCommand` in your SSH Config (`~/.ssh/config`).

On this PR in particular, how would you feel about converting this from taking an `ssh_socks_proxy` to taking something like `additional_ssh_options`.

File: python/ray/autoscaler/updater.py
Comment (vnlitvinov) [reply python/ray/autoscaler/updater.py:281]: According to Python docs, `shlex.join()` is available starting from 3.8 while Ray targets at most 3.6 IIRC.

I thought about using `subprocess.list2cmdline()` originally but then decided to not mix the methods of quoting.

Comment (mehrdadn) [reply python/ray/autoscaler/updater.py:281]: Ahh, thanks for noticing that, I completely forgot! Never mind then :)

**Comment by vnlitvinov** [comment]
@thomasdesr 



> If you're currently blocked on this, you should be able to unblock by setting `ProxyCommand` in your SSH Config (`~/.ssh/config`).

I know and we're currently exploring this option. The drawbacks for this method are that we need to make a blanket rule (apply `ProxyCommand` to all hosts) because it's impossible to predict which name/address a node from, say, AWS would get when it is spawned, so it effectively breaks all internal SSH (which is suboptimal IMHO).



> On this PR in particular, how would you feel about converting this from taking an `ssh_socks_proxy` to taking something like `additional_ssh_options`.

If you look at commit history, that's what I did originally - accepted a whole command to be given to `-o ProxyCommand`. But then I changed my mind because it would make efforts of moving away from using `ssh` as a binary much harder by introducing an extra tight coupling link between the config and the implementation (and I would eventually like to help move using some library for making SSH calls instead, e.g. to https://github.com/ParallelSSH/ssh2-python), as using a library is usually slightly more efficient _and_ it should allow running this without `openssh` (e.g. on native Windows without MSYS).

**Comment by vnlitvinov** [comment]
Putthing proxy option under specific ssh section might make sense, but then I think it would make sense to move other ssh-related options there (like key file), which would make auth section obsolete.
Maybe make a separate `proxy` or `connection` section, which can house `socks` proxy and http/https proxy options? It can house a jump host option, too.

So again, I surely can add a generic "ssh proxy command" option, but that would mean that the final config file would be tied to using OpenSSH _only_ (or would require some Python code to parse and understand this option, which feels like a very dirty hack).

It looks much more clean to specify in the config _what_ needs to be done, not _how_ to do it. Building a command line for `ssh` from those options is trivial, but parsing them out from a user-provided string is not.

I'm not saying you should consider moving to `libssh` straight away, what I'm saying is I want to solve my current problem _and_ at the same time not to hinder potential migration to `libssh`.
That said, this PR is not a step towards migration, but rather I'm trying to not make such migration harder 😉

**Comment by vnlitvinov** [comment]
Note for reviewers: I've rebased off `master` hoping it would fix some MacOS CI failures (as the code that I added should be a no-op unless the newly introduced option is used).

**Comment by thomasdesr** [comment]
I definitely appreciate you taking the broader view :D

I feel like the core of why I'm leaning in the ProxyCommand direction over special casing each alternative transport layer is that OpenSSH was able to defer for (~a decade or so) before they started wanting to add more specialized support. Leaning on a subprocess' stdin & stdout is a pretty durable interface.

On trap door concern, I think we're probably already on the other side. I don't think we can move to another SSH implementation without building support for leaning on ssh_config. I'm sure some users are already leaning on it somewhere :S

**Comment by vnlitvinov** [comment]
> Leaning on a subprocess' stdin & stdout is a pretty durable interface.

While that is true, I personally feel it strange when a _library_ spawns stuff to do its job when it's not necessary, it feels a bit hacky to my taste.



> I don't think we can move to another SSH implementation without building support for leaning on ssh_config

Whan has leaning on `ssh_config` have to have in common with using OpenSSH _binary_? :) The library (if not that one I suggested then some other) could wrap `libssh` correctly and read the config as much as normal `ssh` binary does, so nothing changes from that perspective but it enables other platforms without native ssh (read: Windows).
SSH options present in the config now do not block migration to another platform, but as soon as users are given free reign to add stuff to their configs, this trap's jaws would snap closed. Hence me trying to _not_ fall into this trap with those changes.

Added bonus: my current changes are easier from users' perspective - they don't have to go researching how to pass SSH through proxy, they set a variable with an easily understandable name and off it goes!

**Comment by thomasdesr** [comment]
> > Leaning on a subprocess' stdin & stdout is a pretty durable interface.
> 
> While that is true, I personally feel it strange when a _library_ spawns stuff to do its job when it's not necessary, it feels a bit hacky to my taste.

Completely agree that subprocess is a less than ideal way to build functionality into libraries. That said, this isn't something we're building into our library, it is something OpenSSH exposes to its users to provide alternative transports.

> Whan has leaning on `ssh_config` have to have in common with using OpenSSH _binary_? :) The library (if not that one I suggested then some other) could wrap `libssh` correctly and read the config as much as normal `ssh` binary does, so nothing changes from that perspective but it enables other platforms without native ssh (read: Windows).

I think we're on the same page here :D This is what I was trying to propose.

> SSH options present in the config now do not block migration to another platform, but as soon as users are given free reign to add stuff to their configs, this trap's jaws would snap closed. Hence me trying to _not_ fall into this trap with those changes.

I think we're already past this trap door because we do not know how users' SSH environments are configured today. While we are passing specific options that we need via the exec arguments, that doesn't preclude a user's SSH config from changing any unspecified option to different values. 

Concretely, a user today may be using ProxyCommand/JumpHost to connect to their Ray clusters and relying on a config being set in OpenSSH's ssh_config to make that work.

> Added bonus: my current changes are easier from users' perspective - they don't have to go researching how to pass SSH through proxy, they set a variable with an easily understandable name and off it goes!

Specifically, I'm uncertain about how many users are using SOCKS proxies for SSH transport. It isn't something I've even heard of being in widespread use. What I've personally encountered is more traditional SSH JumpHosts (either via `-o JumpHost` or via netcat) or Google-style SSH-over-http.

Implementing support for ProxyCommand covers both this PR's use case and the common use cases of which I'm aware. So I'm uncertain why we should add special support for this when we can implement the common feature and also gain support for this.

On the discoverability aspect, I expect that most users who have a requirement to use a jump host also have either existing documentation about how to configure SSH w/ OpenSSH's configs, or already have had their clients configured already.

**Comment by vnlitvinov** [comment]
> Concretely, a user today may be using ProxyCommand/JumpHost to connect to their Ray clusters and relying on a config being set in OpenSSH's ssh_config to make that work.

Let's remember we're talking about autoscaler here, which _typically_ is used for a cloud-backed machine pool with seemingly randomly assigned addresses, which makes it rather hard to pre-configure jump hosts or other connection options to, or am I missing something here?

That all said, I surely can revert to an older version of specifying the full `ProxyCommand` in the config file, as I personally don't care much what to put in the config - it's a "set and forget" procedure anyway.

**Comment by thomasdesr** [comment]
I don't disagree that it isn't easy, but I know places that essentially use an SSH config to forward everything in `*.amazonaws.com`. It is a valid use case as much as using a socks proxy.

But I think the core open question is still how do we support other SSH Transports. If more folks show up and say: "Yeah I also use a SOCKS proxy, please make this easier to use, the current solution is too painful" then I'm totally open to revisiting this. I'm just unclear why we should start with a specialized solution when the classic solution will cover both this need and a lot of other user's needs too.

If you are willing going back to a direct proxy_command I'd feel much more comfortable moving this forward.

**Comment by vnlitvinov** [comment]
Okay, I'm finally back to this PR. I've switched back to passing raw `ProxyCommand` value via the config files, and simplified and squashed the commits (after rebasing from current `master`).

**Comment by thomasdesr** [comment]
Looks like other than a few lint issues the rest looks good (flakey tests):

```
./python/ray/autoscaler/updater.py:184:80: E501 line too long (86 > 79 characters)
./python/ray/autoscaler/updater.py:309:22: E128 continuation line under-indented for visual indent
./python/ray/autoscaler/updater.py:319:22: E128 continuation line under-indented for visual indent
```

**Comment by vnlitvinov** [comment]
@thomasdesr linting should be fixed now (at least locally `flake8` sees no more issues).
Also I had to rebase again because #9340 moved the classes I was changing to a separate file 😃

**Review by thomasdesr — APPROVED** [review]

**Comment by thomasdesr** [comment]
Thanks a bunch! I'll get this merged in on Monday :D

**Comment by vnlitvinov** [comment]
@thomasdesr can I ignore CI failures on MacOS (and others apparently caused by those terminating the build)? They don't seem to be in any way related to what I did.

**Comment by thomasdesr** [comment]
@vnlitvinov Yeah, they look unrelated to me too. So I think we can ignore them here X.X

**Comment by vnlitvinov** [comment]
Anything else needed from me to get this merged?

**Comment by thomasdesr** [comment]
Nope, you're good! I'm just tracking down some with with merge button permissions :D

**Comment by vnlitvinov** [comment]
Cool, thanks!

Where can I track a release that will contain this fix? I presume it should be somewhere around 0.8.7, do you have any idea when this would land?

**Comment by thomasdesr** [comment]
If you can use the nightly wheels it should be available soon (I think nightly linux builds have been broken for the last day or two so I'm not sure if it is in the current latest).

`0.8.7` should land sometime in the next 3-5 weeks.

**Pull request merged**
